Added classify method.

primaryobjects opened the pull request:
Added classify method: returns the index of a cluster for a new data point.
Usage:
var clusters = clusterfck.kmeans(colors, 3);
var clusterIndex = clusterfck.kmeans.classify([0, 0, 225]);

Output:
1

primaryobjects committed:
Added classify method: returns the index of a cluster for a new data point.

primaryobjects committed:
Updated README.md

irony commented:
Great work! This is exactly what I'm looking for. Shouldnt the centeroids array be public so it can be serialized and used later when new data should be clustered?

primaryobjects committed:
Changed return type of kmeans() to return the clusters and centroids: { clusters, centroids }.
Added toJSON() method (serializes centroids).
Added fromJSON() method (deserializes centroids).
Added K-Means++ initialization, to replace random initialization of cluster centers.

primaryobjects committed:
Updated readme.

primaryobjects commented:
Actually, I just pushed changes that include toJSON() and fromJSON() methods for serializing the centroids. See https://github.com/primaryobjects/clusterfck

The return value from the kmeans() method now returns an object containing { clusters, centroids }.

I've also added K-means++ initialization of the centroids (as opposed to random initialization).

irony commented:
Great! I will try it out. Maybe to not break backward compatibility it would be better to add the centeroids as a property on the array instead of creating a new object? Like this:

```
var clusters = clusterfck.kmeans(vectors);
console.log(clusters.centroids); 
```

Anyway, good work and perfect timing, I have been using this lib for a while and was just going to expand the lib with this functionality - two days before I did, you pushed the same changes ;)

primaryobjects committed:
Centroids are now publicly accessible. Changed return type of kmeans() method to clusters. Changed return type of fromJSON() to this.

primaryobjects committed:
Re-factored into class.
Use:
var clusterfck = require('clusterfck');
var kmeans = new clusterfck.kmeans();

var clusters = kmeans.train(colors, 3);
var clusterIndex = kmeans.classify(input);

primaryobjects commented:
Hey, give this branch a try. I've converted the functions to a class with prototype methods, so you can now create multiple instances. If it looks good, I'll move it to the master of my fork.
https://github.com/primaryobjects/clusterfck/tree/PROTOTYPE

The new usage would be:
var clusterfck = require('clusterfck');

var kmeans = new clusterfck.kmeans();
var clusters = kmeans.train(colors, 3);
var clusterIndex = kmeans.classify(input);

irony commented:
Cool, i'll let you know!

primaryobjects committed:
Updated readme.

irony commented:
It works fine! I suggest adding this in the constructor?

```
function KMeans(centroids) {
this.centroids = centroids || [];
}
```

That means you can do this:

```
var kmeans = new clusterfck.kmeans(centroids);
```

primaryobjects committed:
Renamed train() method to cluster(). Added optional constructor parameter to accept centroids.

primaryobjects commented:
Done. I've added the optional constructor parameter. Also, renamed train() to cluster() since it makes more sense. 
https://github.com/primaryobjects/clusterfck/commit/9e5cdbdd2e648cf0f2c93f2d808beeec18e7a425

Now, if we can just get harthur to accept the pull request and update the npm package! :)

irony commented:
Great work! Pull it @harthur :+1:

harthur commented:
I'll take a look at this as soon as I can. This is a breaking change, so I'll need to think about it.

irony commented:
It should be possible to do an arguments constructor to make it backwards compatible. It makes the lib much more usable and faster to be able to save the last state when doing reclustering of new data.

harthur commented:
Export the `Kmeans` class so you can do `var kmeans = new clusterfck.Kmeans()`, also export `kmeans` so you can use `var clusters = clusterfck.kmeans(vectors)` as usual, and have that just be like:

``` javascript
exports.kmeans = function(vectors, k) {
   return (new Kmeans()).cluster(vectors, k);
}
```

Save kmeans++ initialization for a follow-up pull request. One thing at a time, unless it's necessary for this.

The indentation in the pull request is a mixed bag of tabs and spaces it looks like. This file is 3-space indented (for better or worse), so stick with that for every line.

I'll take a closer look after these things. Thanks for the PR!

primaryobjects committed:
Added backwards-compatibility for kmeans. Added indentation fixes.
The library can now be used two ways:

var clusters = clusterfck.kmeans(colors, 3);

OR

var kmeans = new clusterfck.Kmeans();
var clusters = kmeans.cluster(colors, 3);

primaryobjects commented:
Done. Added the backwards-compatibility requested and fixed indentation. https://github.com/primaryobjects/clusterfck/commit/d8187230ba2f874826a9ec3bf9c2ad582a3d99a0

The library can now be used two ways:

var clusters = clusterfck.kmeans(colors, 3);

OR

var kmeans = new clusterfck.Kmeans();
var clusters = kmeans.cluster(colors, 3);

harthur commented on README.md:
```diff
@@ -26,10 +27,14 @@ var colors = [
    [255, 13, 8]
 ];
 
-var clusters = clusterfck.kmeans(colors, 3);
```
Keep the basic example here, make another example for using it as a stateful object.

harthur commented on README.md:
````diff
@@ -39,6 +44,49 @@ The second argument to `kmeans` is the number of clusters you want (default is `
 ]
 ```
 
+# Serialization
````
This shouldn't be a main header, it should be under the K-means section as it only pertains to kmeans and not hierarchical clustering.

harthur commented on README.md:
````diff
@@ -39,6 +44,49 @@ The second argument to `kmeans` is the number of clusters you want (default is `
+var clusterIndex = kmeans.classify([0, 0, 225]);
+```
+
+# Initializing with Existing Centroids
````
Same for this header and remaining ones, not a main section.

harthur commented on lib/kmeans.js:
```diff
@@ -8,28 +12,123 @@ function randomCentroids(points, k) {
+}
+  
+KMeans.prototype.initCentroids = function(points, k, distance) {
+   // https://github.com/cmtt/kmeans-js/blob/master/kmeans.js
```
There's no license in that repo, so we can't use it.

primaryobjects replied:
MIT license. https://github.com/cmtt/kmeans-js

harthur commented on lib/kmeans.js:
```diff
@@ -8,28 +12,123 @@ function randomCentroids(points, k) {
+   var addIterator = function (x,y) { return x+y; };
+   var centroids = [];
+
+   /** K-Means++ initialization */
```
Save this for another pull request. Stick to the minimum of changes required for adding a classify method.

harthur commented on lib/kmeans.js:
```diff
@@ -8,28 +12,123 @@ function randomCentroids(points, k) {
+
+   /** K-Means++ initialization */
+
+   /** determine the amount of tries */
```
This looks copied from that other repo, can't do that.

harthur commented on lib/kmeans.js:
```diff
@@ -8,28 +12,123 @@ function randomCentroids(points, k) {
    }
 
-   var centroids = randomCentroids(points, k);
+   this.centroids = this.initCentroids(points, k, distance); // this.randomCentroids(points, k);
```
Use `randomCentroids`. Even so, we'd want to take that commented line out, can always look at the history for previous code.

primaryobjects committed:
Updated README.md. Enabled random initialization as default.

primaryobjects committed:
Update README.md

primaryobjects committed:
Update README.md

primaryobjects commented:
Ok, I've updated the readme to lessen the font on those sub-sections and to bring back the original code example, with a separate one for instantiating the class. I also put back in the random init for centroids. https://github.com/primaryobjects/clusterfck/commit/6c1a0891ca27cb0b2193cc03a258bd42389e8637

Additional edits to readme https://github.com/primaryobjects/clusterfck/commit/7d3626fa9c3453e669eb0930732dce3872eef926

I left in the code for kmeans++ initialization however, as I think it's a valuable enhancement. The original code has an MIT license. I'll leave it active in my own fork, but feel free to accept this pull and then remove it from the master if you prefer.

Let's get it updated on npm! Thanks!

jperl commented:
@primaryobjects Thanks for the kmeans++ I was just about to implement that before finding your repo.

primaryobjects commented:
@jperl I've temporarily put back regular initialization for the pull request for @harthur. Once it's pulled, I'll put the line back in to use kmeans++. Until then, you can grab my fork and just change this line:

https://github.com/primaryobjects/clusterfck/blob/master/lib/kmeans.js#L130

to

``` javascript
// Use kmeans++ initialization.
this.centroids = this.initCentroids(points, k, distance);
```

jperl commented:
@primaryobjects Gotcha, thanks!

irony commented:
What's the status on this. I am still using the separate fork - would be great to be able to intitialize clusterfck with centeroids.

harthur commented:
I merged the code for adding a `classify()` method, and published v0.6.0. Thanks for taking the time to do that primaryobjects.

It's my bad for not making a contribution guide, but in general PRs that are adding multiple features at the same time are harder to review. Even if all the features added are good features (kmeans++ seems like it would be a good addition), you have to determine whether you want all of them at the same time and have to figure out what code goes with what new feature. The complexity makes it easier to put off reviewing it for awhile...

I just marked this repo as unmaintained. You could turn that fork into a new npm package with a new name (retaining the license) and develop from there.

Thanks again.
